File: all/1461_check_if_a_string_contains_all_binary_codes_of_size_k/sol_better_time.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        substrings_showed_list = [False] * (1 << k)
        current_binary_string_value = 0
        binary_string_value_bitmask = ((1 << k) - 1)

        for window_right_index, c in enumerate(s):
            current_binary_string_value = (
                (current_binary_string_value << 1) | (int(c))
            )

            if window_right_index >= (k - 1):
                current_binary_string_value &= binary_string_value_bitmask
                substrings_showed_list[current_binary_string_value] = True

        return all(substrings_showed_list)

File: all/1461_check_if_a_string_contains_all_binary_codes_of_size_k/sol_better_time.c
#include <stdbool.h>
#include <stdlib.h>

bool hasAllCodes(char* s, int k) {
    const int binary_string_value_bitmask = ((1 << k) - 1);

    bool *substrings_showed_array_ptr = (
        (bool*)calloc(1 << k, sizeof(*substrings_showed_array_ptr))
    );
    bool answer = true;
    int window_right_index = 0;
    int current_binary_string_value = 0;

    if (NULL == substrings_showed_array_ptr) {
        return false;
    }

    for (
        int window_right_index = 0;
        '\0' != s[window_right_index];
        ++window_right_index
    ) {
        current_binary_string_value = (
            (current_binary_string_value << 1) | (s[window_right_index] - '0')
        );

        if (window_right_index >= (k - 1)) {
            current_binary_string_value &= binary_string_value_bitmask;
            substrings_showed_array_ptr[current_binary_string_value] = true;
        }
    }

    for (int i = 0; i < (1 << k); ++i) {
        if (!substrings_showed_array_ptr[i]) {
            answer = false;

            break;
        }
    }

    free(substrings_showed_array_ptr);
    substrings_showed_array_ptr = NULL;

    return answer;
}
